wallet-new/test/unit/UTxO/BlockGen.hs

```diff
@@ -0,0 +1,250 @@
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE TemplateHaskell            #-}
+
+module UTxO.BlockGen where
+
+import           Universum hiding (use, (.~))
+
+import           Control.Lens hiding (elements)
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import           Data.Traversable (for)
+import           Pos.Util.Chrono (OldestFirst (..))
+import           Test.QuickCheck
+
+import           UTxO.Bootstrap
+import           UTxO.Context
+import           UTxO.DSL
+import           UTxO.Interpreter
+import           UTxO.PreChain
+import           UTxO.Translate
+
+-- | Blockchain Generator Monad
+--
+-- When generating transactions, we need to keep track of addresses,
+-- balances, and transaction IDs so that we can create valid future
+-- transactions without wasting time searching and validating.
+newtype BlockGen h a
+    = BlockGen
+    { unBlockGen :: StateT (BlockGenCtx h) Gen a
+    } deriving (Functor, Applicative, Monad, MonadState (BlockGenCtx h))
+
+-- | The context and settings for generating arbitrary blockchains.
+data BlockGenCtx h
+    = BlockGenCtx
+    { _bgcAddressesToBalances :: !(Map Addr Value)
+    -- ^ The mapping of current addresses and their current account values.
+    , _bgcFreshHash           :: !Int
+    -- ^ A fresh hash value for each new transaction.
+    , _bgcCurrentBlockchain   :: !(Ledger h Addr)
+    }
+
+makeLenses ''BlockGenCtx
+
+toPreChain
+    :: Hash h Addr
+    => BlockGen h [[Value -> Transaction h Addr]]
+    -> PreChain h Gen
+toPreChain = toPreChainWith identity
+
+toPreChainWith
+    :: Hash h Addr
+    => (BlockGenCtx h -> BlockGenCtx h)
+    -> BlockGen h [[Value -> Transaction h Addr]]
+    -> PreChain h Gen
+toPreChainWith settings bg = PreChain $ \boot -> do
+    ks <- runBlockGenWith settings boot bg
+    return
+        $ OldestFirst . reverse
+        . fmap (OldestFirst . reverse)
+        . zipFees ks
+
+-- | Given an initial bootstrap 'Transaction', this function will
+-- initialize the generator and run the action provided.
+runBlockGen
+    :: Hash h Addr
+    => Transaction h Addr
+    -> BlockGen h a
+    -> Gen a
+runBlockGen = runBlockGenWith identity
+
+-- | Given an initial bootstrap 'Transaction' and a function to customize
+-- the other settings in the 'BlockGenCtx', this function will initialize
+-- the generator and run the action provided.
+runBlockGenWith
+    :: Hash h Addr
+    => (BlockGenCtx h -> BlockGenCtx h)
+    -> Transaction h Addr
+    -> BlockGen h a
+    -> Gen a
+runBlockGenWith settings boot (BlockGen m) =
+    evalStateT m (settings (initializeCtx boot))
```

Comment:
The `settings` attribute here will likely be useful when we expand to more comprehensive tests. We can start with a "seed" blockchain instead of an empty one and ensure that there's not some weird state in that seed that causes a bad transaction failure. Might be overly defensive though.